core/src/main/java/com/linecorp/armeria/common/Flags.java

```diff
@@ -1506,6 +1535,29 @@ private static <T> T getValue(Function<FlagsProvider, @Nullable T> method,
         throw new Error();
     }
 
+    @Nullable
+    private static <T> T getUserValue(Function<FlagsProvider, @Nullable T> method, String flagName) {
```

Comment:
This method looks pretty much the same as the above `getValue` method.
How about removing this method and adding `ignoreDefaultFlags` to the above method instead?
```java
private static <T> T getValue(Function<FlagsProvider, @Nullable T> method,
                              String flagName, boolean ignoreDefaultFlags, Predicate<T> validator) {
    for (FlagsProvider provider : FLAGS_PROVIDERS) {
        if (provider instanceof DefaultFlagsProvider) {
            continue;
        }
...
```

Reply:
I think it's also possible to delegate the `getUserValue` call to `getValue` (so `getValue` calls `getUserValue` and falls back to `DefaultFlagsProvider`)
The reasoning is although it would take up slightly more cpu cycles, I think the method itself is useful.

Reply:
I wonder how you think of `getUserValue` method. Is it a temporary or not temporary? 

I split the method because I thought `getUserValue` method seems to be a temporary method(simply for migration). If it's not temporary, I think it's better to add `ignoreDefaultFlags`. 

ref: https://github.com/line/armeria/pull/5029#discussion_r1278808201

Reply:
> I wonder how you think of getUserValue method. Is it a temporary or not temporary?

Although temporary, we will probably keep the logic for a while (at least until 2.0)

On second thought, would it be simpler to just delete `getUserValue` and modify `DefaultFlagsProvider#useOpenSsl`, `DefaultFlagsProvider#tlsEngineType` to return `null` by default?

Reply:
@jrhee17 That seems to be a great idea 😄

Reply:
@jrhee17 If we delegate `getUserValue` call to `getValue` , `getUserValue`'s `@Nullable` have to be spread to all methods and many more static `Flags` variables which is using `getValue` method(I think the changes are not minor). 
Any idea on how to delegate `@Nullable` `getUserValue` call to non-`@Nullable` `getValue` ?

Reply:
Ah, I see. Then we can probably do
```
if (ignoreDefaultFlags) {
    return null;
}
// Should never reach here because DefaultFlagsProvider always returns a normal value.
throw new Error();
```
at the end.
So it's your call now. We can gather it into one method, which makes the method a little longer, or we can have separate methods. I'm fine with both. It's your call. 😉

Reply:
Ok! I will separate methods because it'll require a lot of changes when `@Nullable` has to be handled by the callers of `getValue` method.

Reply:
>  Any idea on how to delegate @Nullable getUserValue call to non-@Nullable getValue ?

I meant something like the following. Having said this, I think I'm fine with the current approach:

```java
diff --git a/core/src/main/java/com/linecorp/armeria/common/Flags.java b/core/src/main/java/com/linecorp/armeria/common/Flags.java
index ff40a0de6..e30115a4c 100644
--- a/core/src/main/java/com/linecorp/armeria/common/Flags.java
+++ b/core/src/main/java/com/linecorp/armeria/common/Flags.java
@@ -105,7 +105,6 @@ public final class Flags {
                              .sorted(Comparator.comparingInt(FlagsProvider::priority).reversed())
                              .collect(Collectors.toList());
         flagsProviders.add(0, SystemPropertyFlagsProvider.INSTANCE);
-        flagsProviders.add(DefaultFlagsProvider.INSTANCE);
         FLAGS_PROVIDERS = ImmutableList.copyOf(flagsProviders);
     }
 
@@ -561,8 +560,9 @@ public final class Flags {
             return;
         }
 
-        final Boolean useOpenSsl = getUserValue(FlagsProvider::useOpenSsl, "useOpenSsl");
-        final TlsEngineType tlsEngineTypeValue = getUserValue(FlagsProvider::tlsEngineType, "tlsEngineType");
+        final Boolean useOpenSsl = getUserValue(FlagsProvider::useOpenSsl, "useOpenSsl", ignored -> true);
+        final TlsEngineType tlsEngineTypeValue = getUserValue(FlagsProvider::tlsEngineType, "tlsEngineType",
+                                                              ignored -> true);
 
         if (useOpenSsl == null) {
             tlsEngineType = tlsEngineTypeValue != null ? tlsEngineTypeValue : TlsEngineType.OPENSSL;
@@ -1549,28 +1549,16 @@ public final class Flags {
 
     private static <T> T getValue(Function<FlagsProvider, @Nullable T> method,
                                   String flagName, Predicate<T> validator) {
-        for (FlagsProvider provider : FLAGS_PROVIDERS) {
-            try {
-                final T value = method.apply(provider);
-                if (value == null) {
-                    continue;
-                }
-                if (!validator.test(value)) {
-                    logger.warn("{}: {} ({}, validation failed)", flagName, value, provider.name());
-                    continue;
-                }
-                logger.info("{}: {} ({})", flagName, value, provider.name());
-                return value;
-            } catch (Exception ex) {
-                logger.warn("{}: ({}, {})", flagName, provider.name(), ex.getMessage());
-            }
+        final T t = getUserValue(method, flagName, validator);
+        if (t != null) {
+            return t;
         }
-        // Should never reach here because DefaultFlagsProvider always returns a normal value.
-        throw new Error();
+        return method.apply(DefaultFlagsProvider.INSTANCE);
     }
 
     @Nullable
-    private static <T> T getUserValue(Function<FlagsProvider, @Nullable T> method, String flagName) {
+    private static <T> T getUserValue(Function<FlagsProvider, @Nullable T> method, String flagName,
+                                      Predicate<T> validator) {
         for (FlagsProvider provider : FLAGS_PROVIDERS) {
             if (provider instanceof DefaultFlagsProvider) {
                 continue;
@@ -1581,6 +1569,10 @@ public final class Flags {
                 if (value == null) {
                     continue;
                 }
+                if (!validator.test(value)) {
+                    logger.warn("{}: {} ({}, validation failed)", flagName, value, provider.name());
+                    continue;
+                }
 
                 logger.info("{}: {} ({})", flagName, value, provider.name());
                 return value;

```

Reply:
Oh, thanks for the great detail. I think above approach looks much clean. Let me apply above first and see whether no tests fail.

Reply:
@jrhee17
I think calling `getUserValue` from `getValue` adds some overhead because of `provider instanceof DefaultFlagsProvider`(even though we don't need except for `useOpenSsl` and `tlsEngineType`). But I think the overheads are small, so I'll go with the one you recommended.